### ISSUE TITLE
Incorrect typing when calling splice() on a tuple

### ISSUE THREAD

**Issue by bbrk24** [microsoft/TypeScript#45355]
# Bug Report

### 🔎 Search Terms

tuple splice

### 🕗 Version & Regression Information

- This is the behavior in every version I tried. I reviewed the FAQ for entries about tuples, but didn't find anything.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0#code/CYUwxgNghgTiAEEQBd6xgLngbQM7JgEsA7AcwBp5iBXAWwCMQYBdAbgFgAoL9AOlwAOEQmBAAKAIwBKVvAD0c+LWrIoyELjQwYlWlADWJUvEKoA9sQgBPHPiJlmXJKnQAmeAF4tMWQpPEwM21wZGt-ADMmOGA0TTwCI0oaBiZHTmdvAGZPeAAeePtSZgA+dF9FEkDgsFRA4gA3JlxCC3gooK4uOQAqeABJeDhA0mJCAC8EZAALNRNkAHJNYjMXCAB3KCtNacJNEChcG2QzeFwQBDWZ1GQrAUmphCDCUhIoCC54XuRqIQRdpRUahAMWOvHgAAUmDMBJpTPBAtQIDFGIFaAgoMQrNhmPAgvBqMR9Ms1sRsZQ8dQzh9elB8aNWjc7vAxCR4DtNGADiBKGI7EZ4AAfKh0RgwKTYqRggDyMCUm0Y8LewjI1NOQhEj2IaDZPyQpymZkRyL+ECQpDe-jZD3g4UIMHw8CEUFEYL6WoxNk5Z0o0xAqrA1GCxFQjBm9Rasv+lSCQ2QvGpciAA)

### 💻 Code

```ts
declare let arr: [string, number];

arr.splice(1); // mutates arr, making it only [string]
let arr2 = arr; // incorrectly inferred as [string, number]
let arr3 = arr as [string]; // incorrect conversion error
```

### 🙁 Actual behavior

See the comments in code above. What the TypeScript compiler infers is different from the actual runtime behavior.  
It's worth noting that the return value of the call to `splice()` is correctly inferred as `(string | number)[]`.

### 🙂 Expected behavior

Either calling `arr.splice(1)` should be an error (due to calling splice on a tuple), or the type of `arr` should be different after calling `arr.splice(1)`.

**Comment by MartinJohns** [comment]
Related, or perhaps even a duplicate of #40316.

**Comment by bbrk24** [comment]
That's true, I didn't see that. The same problem exists with `shift()` as well.
